File: elixir/lib/data_structures/binary_search_tree.ex
defmodule DataStructures.BinarySearchTree do
  @moduledoc """
  Binary search tree implementation.
  Time: O(log n) average, O(n) worst case
  Space: O(n)
  """

  defstruct value: nil, left: nil, right: nil, count: 0

  @type t :: %__MODULE__{
          value: any(),
          left: t() | nil,
          right: t() | nil,
          count: non_neg_integer()
        }

  @doc "Create an empty tree"
  @spec new() :: t()
  def new, do: %__MODULE__{}

  @doc "Create from list"
  @spec from_list(list()) :: t()
  def from_list(list) do
    Enum.reduce(list, new(), &insert(&2, &1))
  end

  @doc "Insert a value"
  @spec insert(t(), any()) :: t()
  def insert(%__MODULE__{count: 0}, value) do
    %__MODULE__{value: value, count: 1}
  end

  def insert(%__MODULE__{value: node_value, left: left, right: right, count: count} = tree, value) do
    cond do
      value < node_value ->
        %{tree | left: insert(left || new(), value), count: count + 1}

      value > node_value ->
        %{tree | right: insert(right || new(), value), count: count + 1}

      true ->
        %{tree | value: value}
    end
  end

  @doc "Check if value exists"
  @spec contains?(t(), any()) :: boolean()
  def contains?(%__MODULE__{count: 0}, _value), do: false

  def contains?(%__MODULE__{value: node_value, left: left, right: right}, value) do
    cond do
      value < node_value -> contains?(left || new(), value)
      value > node_value -> contains?(right || new(), value)
      true -> true
    end
  end

  @doc "Remove a value"
  @spec remove(t(), any()) :: t()
  def remove(%__MODULE__{count: 0} = tree, _value), do: tree

  def remove(%__MODULE__{value: node_value, left: left, right: right} = tree, value) do
    cond do
      value < node_value ->
        new_left = if left, do: remove(left, value), else: nil
        new_count = count_nodes(new_left) + count_nodes(right) + 1
        %{tree | left: new_left, count: new_count}

      value > node_value ->
        new_right = if right, do: remove(right, value), else: nil
        new_count = count_nodes(left) + count_nodes(new_right) + 1
        %{tree | right: new_right, count: new_count}

      true ->
        remove_node(tree)
    end
  end

  defp remove_node(%__MODULE__{left: nil, right: nil}), do: new()
  defp remove_node(%__MODULE__{left: nil, right: right}), do: right
  defp remove_node(%__MODULE__{left: left, right: nil}), do: left

  defp remove_node(%__MODULE__{left: left, right: right}) do
    {:ok, successor} = find_min(right)
    new_right = remove(right, successor)
    new_count = count_nodes(left) + count_nodes(new_right) + 1
    %__MODULE__{value: successor, left: left, right: new_right, count: new_count}
  end

  defp count_nodes(nil), do: 0
  defp count_nodes(%__MODULE__{count: count}), do: count

  @doc "Find minimum value"
  @spec find_min(t()) :: {:ok, any()} | {:error, :empty}
  def find_min(%__MODULE__{count: 0}), do: {:error, :empty}
  def find_min(%__MODULE__{value: value, left: nil}), do: {:ok, value}
  def find_min(%__MODULE__{left: left}), do: find_min(left)

  @doc "Find maximum value"
  @spec find_max(t()) :: {:ok, any()} | {:error, :empty}
  def find_max(%__MODULE__{count: 0}), do: {:error, :empty}
  def find_max(%__MODULE__{value: value, right: nil}), do: {:ok, value}
  def find_max(%__MODULE__{right: right}), do: find_max(right)

  @doc "Check if tree is empty"
  @spec empty?(t()) :: boolean()
  def empty?(%__MODULE__{count: 0}), do: true
  def empty?(%__MODULE__{}), do: false

  @doc "Get number of nodes"
  @spec size(t()) :: non_neg_integer()
  def size(%__MODULE__{count: count}), do: count

  @doc "Inorder traversal (sorted order)"
  @spec inorder(t()) :: list()
  def inorder(%__MODULE__{count: 0}), do: []

  def inorder(%__MODULE__{value: value, left: left, right: right}) do
    inorder(left || new()) ++ [value] ++ inorder(right || new())
  end

  @doc "Preorder traversal"
  @spec preorder(t()) :: list()
  def preorder(%__MODULE__{count: 0}), do: []

  def preorder(%__MODULE__{value: value, left: left, right: right}) do
    [value] ++ preorder(left || new()) ++ preorder(right || new())
  end

  @doc "Postorder traversal"
  @spec postorder(t()) :: list()
  def postorder(%__MODULE__{count: 0}), do: []

  def postorder(%__MODULE__{value: value, left: left, right: right}) do
    postorder(left || new()) ++ postorder(right || new()) ++ [value]
  end

  @doc "Level order traversal (breadth-first)"
  @spec level_order(t()) :: list()
  def level_order(%__MODULE__{count: 0}), do: []

  def level_order(tree) do
    do_level_order([tree], [])
  end

  defp do_level_order([], acc), do: Enum.reverse(acc)

  defp do_level_order([%__MODULE__{count: 0} | rest], acc) do
    do_level_order(rest, acc)
  end

  defp do_level_order([%__MODULE__{value: value, left: left, right: right} | rest], acc) do
    children =
      [left, right]
      |> Enum.filter(&(&1 != nil && &1.count > 0))

    do_level_order(rest ++ children, [value | acc])
  end

  @doc "Convert to sorted list"
  @spec to_list(t()) :: list()
  def to_list(tree), do: inorder(tree)
end

File: zig/src/data_structures/binary_search_tree.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

/// Binary search tree with standard operations.
/// Time: O(log n) average, O(n) worst case
/// Space: O(n)
pub fn BinarySearchTree(comptime T: type) type {
    return struct {
        const Self = @This();
        const Node = struct {
            value: T,
            left: ?*Node,
            right: ?*Node,
        };

        allocator: Allocator,
        root: ?*Node,
        count: usize,

        pub fn init(allocator: Allocator) Self {
            return .{
                .allocator = allocator,
                .root = null,
                .count = 0,
            };
        }

        pub fn deinit(self: *Self) void {
            self.freeNode(self.root);
        }

        fn freeNode(self: *Self, node: ?*Node) void {
            if (node) |n| {
                self.freeNode(n.left);
                self.freeNode(n.right);
                self.allocator.destroy(n);
            }
        }

        pub fn insert(self: *Self, value: T) !void {
            self.root = try self.insertNode(self.root, value);
        }

        fn insertNode(self: *Self, node: ?*Node, value: T) !?*Node {
            if (node == null) {
                const new_node = try self.allocator.create(Node);
                new_node.* = .{ .value = value, .left = null, .right = null };
                self.count += 1;
                return new_node;
            }

            const n = node.?;
            if (value < n.value) {
                n.left = try self.insertNode(n.left, value);
            } else if (value > n.value) {
                n.right = try self.insertNode(n.right, value);
            } else {
                n.value = value;
            }
            return n;
        }

        pub fn contains(self: *const Self, value: T) bool {
            return self.findNode(self.root, value) != null;
        }

        fn findNode(self: *const Self, node: ?*Node, value: T) ?*Node {
            _ = self;
            if (node == null) return null;
            const n = node.?;
            if (value < n.value) return self.findNode(n.left, value);
            if (value > n.value) return self.findNode(n.right, value);
            return n;
        }

        pub fn findMin(self: *const Self) ?T {
            if (self.root == null) return null;
            var current = self.root;
            while (current.?.left) |left| {
                current = left;
            }
            return current.?.value;
        }

        pub fn findMax(self: *const Self) ?T {
            if (self.root == null) return null;
            var current = self.root;
            while (current.?.right) |right| {
                current = right;
            }
            return current.?.value;
        }

        pub fn isEmpty(self: *const Self) bool {
            return self.count == 0;
        }

        pub fn size(self: *const Self) usize {
            return self.count;
        }

        pub fn inorder(self: *const Self, allocator: Allocator) !std.ArrayList(T) {
            var result = std.ArrayList(T).init(allocator);
            try self.inorderTraversal(self.root, &result);
            return result;
        }

        fn inorderTraversal(self: *const Self, node: ?*Node, result: *std.ArrayList(T)) !void {
            _ = self;
            if (node) |n| {
                try self.inorderTraversal(n.left, result);
                try result.append(n.value);
                try self.inorderTraversal(n.right, result);
            }
        }
    };
}

test "BinarySearchTree operations" {
    const allocator = std.testing.allocator;
    var bst = BinarySearchTree(i32).init(allocator);
    defer bst.deinit();

    try bst.insert(5);
    try bst.insert(3);
    try bst.insert(7);
    try bst.insert(1);
    try bst.insert(9);

    try std.testing.expect(bst.contains(5));
    try std.testing.expect(bst.contains(3));
    try std.testing.expect(!bst.contains(10));

    try std.testing.expectEqual(@as(?i32, 1), bst.findMin());
    try std.testing.expectEqual(@as(?i32, 9), bst.findMax());
}
